File: semantic_layer/ood.py
from __future__ import annotations

from typing import Any, Dict

import numpy as np
from numpy.typing import NDArray


class OODDetector:
    """
    Diagonal Mahalanobis distance using Welford's online mean/variance.
    Cold start: returns 0.0 until min_samples reached (no OOD blocking).
    """

    def __init__(self, dim: int, min_samples: int = 50) -> None:
        self.dim = dim
        self.min_samples = min_samples
        self._count = 0
        self._mean = np.zeros(dim, dtype=np.float64)
        # Welford's aggregate sum of squared deviations. MUST start at zero;
        # seeding it with ones inflated every dimension's variance (~9x for
        # L2-normalized MiniLM embeddings), which deflated the Mahalanobis
        # distance ~3x and effectively disabled the strict-OOD path while the
        # detector was young. The `ready` gate (count >= min_samples) plus the
        # epsilon floor in `score` already prevent divide-by-zero.
        self._m2 = np.zeros(dim, dtype=np.float64)

    def update(self, embedding: NDArray[np.float32]) -> None:
        self._count += 1
        x = embedding.astype(np.float64)
        delta = x - self._mean
        self._mean += delta / self._count
        delta2 = x - self._mean
        self._m2 += delta * delta2

    @property
    def ready(self) -> bool:
        return self._count >= self.min_samples

    @property
    def sample_count(self) -> int:
        return self._count

    def score(self, embedding: NDArray[np.float32]) -> float:
        if not self.ready:
            return 0.0
        variance = self._m2 / max(self._count - 1, 1)
        diff = embedding.astype(np.float64) - self._mean
        return float(np.sqrt(np.sum((diff**2) / (variance + 1e-6))))

    def is_ood(self, embedding: NDArray[np.float32], threshold: float) -> bool:
        return self.score(embedding) > threshold

    def state(self) -> Dict[str, Any]:
        """Serializable snapshot for persistence across restarts."""
        return {
            "dim": self.dim,
            "min_samples": self.min_samples,
            "count": self._count,
            "mean": self._mean.tolist(),
            "m2": self._m2.tolist(),
        }

    def load_state(self, state: Dict[str, Any]) -> bool:
        """Restore a previously saved snapshot. Ignores dimension mismatches."""
        mean = np.asarray(state.get("mean", []), dtype=np.float64)
        m2 = np.asarray(state.get("m2", []), dtype=np.float64)
        if mean.shape != (self.dim,) or m2.shape != (self.dim,):
            return False
        self._count = int(state.get("count", 0))
        self._mean = mean
        self._m2 = m2
        return True

File: src-tauri/src/semantic_layer/ood.rs
//! Online out-of-distribution detection via diagonal Mahalanobis distance.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OodState {
    pub dim: usize,
    pub min_samples: usize,
    pub count: usize,
    pub mean: Vec<f64>,
    pub m2: Vec<f64>,
}

pub struct OodDetector {
    dim: usize,
    pub min_samples: usize,
    count: usize,
    mean: Vec<f64>,
    m2: Vec<f64>,
}

impl OodDetector {
    pub fn new(dim: usize, min_samples: usize) -> Self {
        Self {
            dim,
            min_samples,
            count: 0,
            mean: vec![0.0; dim],
            m2: vec![0.0; dim],
        }
    }

    pub fn update(&mut self, embedding: &[f32]) {
        if embedding.len() != self.dim {
            return;
        }
        self.count += 1;
        for i in 0..self.dim {
            let x = embedding[i] as f64;
            let delta = x - self.mean[i];
            self.mean[i] += delta / self.count as f64;
            let delta2 = x - self.mean[i];
            self.m2[i] += delta * delta2;
        }
    }

    pub fn ready(&self) -> bool {
        self.count >= self.min_samples
    }

    pub fn sample_count(&self) -> usize {
        self.count
    }

    pub fn score(&self, embedding: &[f32]) -> f32 {
        if !self.ready() || embedding.len() != self.dim {
            return 0.0;
        }
        let mut sum = 0.0f64;
        for i in 0..self.dim {
            let variance = self.m2[i] / (self.count.saturating_sub(1).max(1) as f64);
            let diff = embedding[i] as f64 - self.mean[i];
            sum += (diff * diff) / (variance + 1e-6);
        }
        sum.sqrt() as f32
    }

    pub fn state(&self) -> OodState {
        OodState {
            dim: self.dim,
            min_samples: self.min_samples,
            count: self.count,
            mean: self.mean.clone(),
            m2: self.m2.clone(),
        }
    }

    pub fn load_state(&mut self, state: &OodState) -> bool {
        if state.mean.len() != self.dim || state.m2.len() != self.dim {
            return false;
        }
        self.count = state.count;
        self.mean = state.mean.clone();
        self.m2 = state.m2.clone();
        true
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn cold_start_returns_zero_score() {
        let detector = OodDetector::new(8, 5);
        let emb = vec![1.0 / (8.0f32).sqrt(); 8];
        assert_eq!(detector.score(&emb), 0.0);
        assert!(!detector.ready());
    }

    #[test]
    fn state_round_trip() {
        let mut detector = OodDetector::new(8, 5);
        for seed in 0..20 {
            let emb: Vec<f32> = (0..8)
                .map(|i| ((seed + i) as f32 * 0.01).sin())
                .collect();
            detector.update(&emb);
        }
        let state = detector.state();
        let mut restored = OodDetector::new(8, 5);
        assert!(restored.load_state(&state));
        let probe = vec![1.0; 8];
        assert_eq!(restored.sample_count(), detector.sample_count());
        assert!((restored.score(&probe) - detector.score(&probe)).abs() < 1e-6);
    }
}
